File: haskell/bob/Bob.hs
{-# LANGUAGE OverloadedStrings #-}
module Bob (responseFor) where
import Data.Char (toUpper)
import qualified Data.Text as T (length, strip, pack)
import Text.Regex.Posix ((=~))

responseFor :: String -> String
responseFor phrase
  | T.length (T.strip $ T.pack phrase) == 0 = "Fine. Be that way!"
  | shouting phrase = "Whoa, chill out!"
  | head (reverse phrase) == '?' = "Sure."
  | otherwise = "Whatever."
  where shouting p = map toUpper p == p &&
          p =~ ("[A-Z]" :: String) :: Bool

File: scala/bob/src/main/scala/Bob.scala
/**
  * Created by jgaylard on 21/04/17.
  */
class Bob {
  def hey(s: String) : String = s match {
    case s if (s == s.toUpperCase())
      && s.exists(_.isLetter)=> "Whoa, chill out!"
    case s if s.endsWith("?") => "Sure."
    case s if s.trim.isEmpty => "Fine. Be that way!"
    case _ => "Whatever."
  }
}
